python/ml4ir/applications/ranking/features/feature_fns/string.py

```diff
@@ -63,4 +65,96 @@ def call(self, inputs, training=None):
             query_len = tf.strings.length(inputs)
 
         query_len = tf.expand_dims(tf.cast(query_len, tf.float32), axis=-1)
-        return query_len
\ No newline at end of file
+        return query_len
+
+
+class QueryTypeVector(BaseFeatureLayerOp):
+    """
+    Compute the length of the query string context feature
+    """
+    LAYER_NAME = "query_type_vector"
+
+    REMOVE_QUOTES = "remove_quotes"
+    OUTPUT_MODE = "output_mode"
+    EMBEDDING_SIZE = "embedding_size"
+
+    ALPHA_QUERY_TYPE = "alpha"
+    NUMERIC_QUERY_TYPE = "num"
+    PUNCTUATION_QUERY_TYPE = "punct"
+
+    ALPHA_REGEX = r"^.*[a-zA-Z]+.*$"
+    NUMERIC_REGEX = r"^.*[0-9]+.*$"
+    PUNCTUATION_REGEX = "^.*[" + "".join([re.escape(c) for c in list(string.punctuation)]) + "]+.*$"
+
+
+    VOCABULARY = [
+        ALPHA_QUERY_TYPE,
+        NUMERIC_QUERY_TYPE,
+        PUNCTUATION_QUERY_TYPE,
+        ALPHA_QUERY_TYPE + NUMERIC_QUERY_TYPE,
+        ALPHA_QUERY_TYPE + PUNCTUATION_QUERY_TYPE,
+        NUMERIC_QUERY_TYPE + PUNCTUATION_QUERY_TYPE,
+        ALPHA_QUERY_TYPE + NUMERIC_QUERY_TYPE + PUNCTUATION_QUERY_TYPE
+    ]
```

Comment:
This seemed like the easiest and cleanest way to implement this logic in tensorflow. Let me know if you have ideas. The way I think about it - this is a naive v0 version of query type we use to quickly get this feature to prod. In the near future, we might want to explore a separate embedding learned from a different/multiple tasks. Can discuss more offline.

For why the vocab looks like this -the query type today is generated like this -
<contain alphabets?> + <contains numbers?> + <contains punctuations?>

This leads to 2 * 2 * 2 total concatenated "types". You see only 7 here because the "" query type is handled by default as the out of vocabulary vector.

Reply:
gotcha